### ISSUE TITLE
[Feature Request] Import from PuTTY/KiTTY/etc

### ISSUE THREAD

**Issue by valdearg** [xpipe-io/xpipe#342]
Wondering if it might be possible to support importing from PuTTY/KiTTY (KiTTY being a fork of PuTTY).

The settings are stored within the registry at the below locations:

KiTTY: HKEY_CURRENT_USER\Software\9bis.com\KiTTY\Sessions

PuTTY: HKEY_CURRENT_USER\Software\SimonTatham\PuTTY\Sessions

This looks like the following in the regedit:

![image](https://github.com/user-attachments/assets/2078cc39-442b-400f-a73c-35db85948f76)

Fields of interest may be the:

HostName - connection hostname
UserName - username to pass to the SSH session.
Protocol - with PuTTY supporting multiple protocols, this can be ssh, telnet, serial, etc.
Autocommand - this is a command which is sent to the SSH session once logged in. 

The session name would come from the registry key name, it seems to use HTML encoding so %20 as spaces etc.

**Comment by crschnick** [comment]
Hmm, there is still some information missing like authentication methods such as key files. Such an import functionality would only result in incomplete connections.

Do these tools support exporting to a SSH config file?

**Comment by valdearg** [comment]
Thanks for taking a look! There's a couple of options stored in the registry key regarding authentication:

Password: this seems to be an encrypted value
PublicKeyFile: this is the file path to a putty version of the public key, this comes with an extension .ppk

There is then a key for AgentFwd which means "Allow agent forwarding". PuTTY has a version of the SSH agent where the keys can be loaded into, then used globally.

Looks like there's one project here: https://github.com/abdrakhmanov/PuTTY-to-OpenSSH-Converter might be an option.

**Comment by crschnick** [comment]
Yeah the password can probably not be decrypted. The key files are in a putty-specific format and there are no command-line tools from putty available to convert that into a normal .pem key on Windows. (I guess that's vendor lock in for you) XPipe already support the putty agent (pageant).

**Comment by crschnick** [comment]
Yeah without being able to convert keys I think this is going to be difficult

**Comment by valdearg** [comment]
I wonder if making a little popup warning might be an option, something like "Please note: passwords and SSH key configs from PuTTY are not imported, these should be updated manually".

Then importing the configs themselves.

**Comment by crschnick** [comment]
But then you still have to go through every single key in putty and export it manually and then add it in XPipe.

**Comment by valdearg** [comment]
Good point there, if someone has lots it would be a pain. 

Another possibility might be adding in support for the plink application, it can load sessions with the `-load` option and it can also take the key using `-i` options.

**Comment by crschnick** [comment]
In theory it could just run `plink -load <name>`, yes. But that would not really be an import, more like another integration for a tool.

**Comment by seanishereforthefood** [comment]
> But then you still have to go through every single key in putty and export it manually and then add it in XPipe.

I would rather have the option to import incomplete sessions as that's a start versus creating everything manually. It would be best if functionality to import similar to MobaXTerm and Termius is considered as a feature.

**Comment by crschnick** [comment]
I mean effectively it would only really import the connection name and host address in that scenario. The rest could either not be imported, like the keys, or could not be mapped as many options are completely different between xpipe and putty. The same would probably go for mobaxterm. With termius, I don't think there will be an import functionality because their vault is completely locked down not accessible to any outside tool.

**Comment by seanishereforthefood** [comment]
> I mean effectively it would only really import the connection name and host address in that scenario. The rest could either not be imported, like the keys, or could not be mapped as many options are completely different between xpipe and putty. The same would probably go for mobaxterm. With termius, I don't think there will be an import functionality because their vault is completely locked down not accessible to any outside tool.

If only name and address are imported, that's still better than nothing if one manages dozens or hundreds of hosts. While Termius does have an import function (from MobaXTerm, PuTTY, .csv, SecureCRT, ssh_config), it's lacking an export function, which is what drove me to look into Xpipe; however, importing is blocking me from considering this at all. 

![Image](https://github.com/user-attachments/assets/db310d6a-7f41-4f6c-b672-a515d4f47a88)

**Comment by crschnick** [comment]
XPipe has a fully supported import functionality for OpenSSH config files, that is the only format that is actually usable. If other tools could export to that, would be a whole lot easier.

But yes I see your point. I can look into adding some port of import functionality, although that one will still be incomplete. If that is still satisfactory, then I guess it works. I was holding off on this because that didn't seem that desirable

**Comment by crschnick** [comment]
So I worked on this in https://github.com/xpipe-io/xpipe-ptb/releases

If you want, you can quickly try whether the import works correctly. The rest of the build is pretty unstable, so I wouldn't recommend using that other than for testing this.
